src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5594,3 +5600,47 @@ ExtractParametersFromParamList(ParamListInfo paramListInfo,
 																   parameterData->value);
 	}
 }
+
+
+/*
+ * Delegated functions which are not part of another transaction are naturally safe,
+ * but if it's part of an outer transaction (a 2PC), we should ensure that the current
+ * task will not do remote(connection) activity to ensure the transactional integrity
+ * of the current participating 2PC.
+ */
+static void
+EnsureDelegatedFunctionSafeExecution(List *taskList)
+{
+	if (taskList == NIL)
+	{
+		/* we allow executions with an empty task list (e.g. WHERE false) */
+		return;
+	}
+
+	if (!IsCitusInitiatedRemoteBackend())
+	{
+		/* application-initiated backends coordinate the 2PC, need no restrictions */
```

Comment:
maybe: /* we are in the client-initiated backend, remote tasks are ok */

Reply:
Changed